python/requirements.txt

```diff
@@ -59,5 +58,6 @@ typer
 fsspec
 pandas>=1.3
 pydantic<2
+pynvml>=11.5.0
```

Comment:
[pynvml](https://github.com/gpuopenanalytics/pynvml) should be avoided, one should use [nvidia-ml-py](https://pypi.org/project/nvidia-ml-py/).

Reply:
@wookayin do you know the difference between these two?

Reply:
`nvidia-ml-py` is the only official python NVML bindings maintained by NVIDIA. `pynvml` is not. I have no idea why this package exists, only adds a confusion. It adds some additional packages but the core nvml part is simply a copy, redistribution of the official pynvml. It's also not up-to-date (as of now, it doesn't support R535 drivers for instance, although the use here would be still backward-compatible), and used to behave *differently* than the official one and has caused some conflicts in the past.

Reply:
@wookayin Thanks for the detailed explanation! 

> It's also not up-to-date (as of now, it doesn't support R535 drivers for instance, although the use here would be still backward-compatible)

Since we use pynvml to auto detect the number of gpus, will an old version of pynvml be able to detect newer version of gpus or users need to use the latest nvidia-ml-py.

Reply:
That's true, for this purpose needed by ray, either will work. But other packages (including gpustat and nvitop to name a few) requires nvidia-ml-py, and if both packages (`pip install pynvml nvidia-ml-py`) are installed, which provides the same python package, then one will overwrite the other, which will break the `pynvml` package. In other words, installing `ray` shipping with the wrong dependency might break `gpustat` or `nvitop`. This should be avoided, so `nvidia-ml-py` should be used instead of `pynvml`.

Reply:
Yea we will use `nvidia-ml-py` not `pynvml`. But instead of depending on `nvidia-ml-py` in setup.py, we are considering just copying this `pynvml.py` file into Ray codebase. Do you see a problem with this approach? Like will, let's say, nvidia releases a new gpu tomorrow and the existing `pynvml.py` won't be able to auto-detect the count.

Reply:
NVIDIA's pynvml often introduces unexpected breaking changes -- but in the past they were mostly about the process information. It was more than once so I don't fully trust NVIDIA for the guaranteed backward compatibility. That being said, it's quite unlikely the API for # of gpus, or memory utilization will break, but there's no guarantee. I think bundling `pynvml` is also a good idea which will be more protective against potential breaking changes -- as long as maintaining it is not a problem.

Reply:
Sg. I think then bundling `pynvml` is the way to go for us.